keps/sig-storage/20190120-execution-hook-design.md

```diff
@@ -0,0 +1,520 @@
+        // all pods and containers based on PodContainerSelector.
+        // If both PodContainerNamesList and PodContainerSelector are not
+        // specified, the ExecutionHook cannot be executed and it will fail.
+        // +optional
+        PodContainerNamesList []PodContainerNames
+
+        // PodContainerSelector is for hook controller to find pods and containers
+        // based on the pod label selector and container names
+        // If PodContainerNamesList is specified, this field will not be used.
+        // +optional
+        PodContainerSelector *podContaienrSelector
+}
+
+type PodContainerNames struct {
+        PodName string
+        ContainerNames []string
+}
+
+type PodContainerSelector struct {
+	// PodSelector specifies a label query over a set of pods.
+        // +optional
+	PodSelector *metav1.LabelSelector
+
+        // ContainerSelector is a selector which must be true for the hook to run on the container
+        // Hook controller will find all containers in the selected pods based on ContainerSelector,
+        // if specified.
+        // The ContainerSeletor maps pod label to container name(s) so that the controller only select
+	// the containers that are listed in the map. Otherwise, all containers of the pods that match
+	// the pod label selector will be selected.
+        // +optional
+        ContainerSelector map[string]string
+ }
+```
+
+Here is the definition of the ExecutionHookStatus. This represents the current state of a hook for all selected containers in the selected pods.
+
+```
+// ExecutionHookStatus represents the current state of a hook
+type ExecutionHookStatus struct {
+        // This is a list of ContainerExecutionHookStatus, with each status representing information
+        // about how hook is executed in a container, including pod name, container name, ActionTimestamp,
+        // ActionSucceed, etc.
+        // +optional
+        ContainerExecutionHookStatuses []ContainerExecutionHookStatus
+
+        // Action Summary status
+        // Default is nil
+        // +optional
+        ActionSucceed *bool
+}
+
+// ContainerExecutionHookStatus represents the current state of a hook for a specific container in a pod
+type ContainerExecutionHookStatus struct {
+        // This field is required
+        PodName string
+
+        // This field is required
+        ContainerName string
+
+        ActionStatus *ExecutionHookActionStatus
+ }
+
+type ExecutionHookActionStatus struct {
+        // If not set, it is nil, indicating Action has not started
+        // If set, it means Action has started at the specified time
+	// +optional
+        ActionTimestamp *int64
+
+        // If not set, it is nil, indicating Action is not complete on the specified container in the pod
+        // +optional
+        ActionSucceed *bool
+
+        // The last error encountered when running the action
+	// +optional
+        Error *HookError
+}
+
+type HookError storage.VolumeError
```

Comment:
I think we can simply make a copy of `VolumeError` if it's the right type to use here, to avoid a dependency on `storage`.

Reply:
we are going to define our own error here to support different errortype

Reply:
Sounds good.